src/main/scala/com/microsoft/ml/spark/lightgbm/LightGBMBooster.scala

```diff
@@ -28,36 +28,72 @@ class LightGBMBooster(val model: String) extends Serializable {
       if (raw) lightgbmlibConstants.C_API_PREDICT_RAW_SCORE
       else lightgbmlibConstants.C_API_PREDICT_NORMAL
     features match {
-      case dense: DenseVector => predictForMat(dense.toArray, kind, classification)
-      case sparse: SparseVector => predictForCSR(sparse, kind, classification)
+      case dense: DenseVector => predictScoreForMat(dense.toArray, kind, classification)
+      case sparse: SparseVector => predictScoreForCSR(sparse, kind, classification)
+    }
+  }
+
+  def predictLeaf(features: Vector): Array[Double] = {
+    // Reload booster on each node
+    if (boosterPtr == null) {
+      LightGBMUtils.initializeNativeLibrary()
+      boosterPtr = getBoosterPtrFromModelString(model)
+    }
+
+    features match {
+      case dense: DenseVector => predictLeafForMat(dense.toArray)
+      case sparse: SparseVector => predictLeafForCSR(sparse)
     }
   }
 
   lazy val numClasses: Int = getNumClasses()
 
+  lazy val numTotalModel: Int = getNumTotalModel
+
+  lazy val numModelPerIteration: Int = getNumModelPerIteration
+
+  lazy val numIterations: Int = numTotalModel / numModelPerIteration
+
   @transient
   var scoredDataOutPtr: SWIGTYPE_p_double = _
 
   @transient
   var scoredDataLengthLongPtr: SWIGTYPE_p_long_long = _
 
+  @transient
+  var leafIndexDataOutPtr: SWIGTYPE_p_double = _
+
+  @transient
+  var leafIndexDataLengthLongPtr: SWIGTYPE_p_long_long = _
+
   def ensureScoredDataCreated(): Unit = {
-    if (scoredDataLengthLongPtr != null)
-      return
+    if (scoredDataLengthLongPtr == null) {
+      scoredDataOutPtr = lightgbmlib.new_doubleArray(numClasses)
+      scoredDataLengthLongPtr = lightgbmlib.new_int64_tp()
+      lightgbmlib.int64_tp_assign(scoredDataLengthLongPtr, 1)
+    }
+  }
 
-    scoredDataOutPtr = lightgbmlib.new_doubleArray(numClasses)
-    scoredDataLengthLongPtr = lightgbmlib.new_int64_tp()
-    lightgbmlib.int64_tp_assign(scoredDataLengthLongPtr, 1)
+  def ensureLeafIndexDataCreated(): Unit = {
+    if (leafIndexDataOutPtr == null) {
+      leafIndexDataOutPtr = lightgbmlib.new_doubleArray(numTotalModel)
+      leafIndexDataLengthLongPtr = lightgbmlib.new_int64_tp()
+      lightgbmlib.int64_tp_assign(leafIndexDataLengthLongPtr, numTotalModel)
+    }
   }
 
   override protected def finalize(): Unit = {
     if (scoredDataLengthLongPtr != null)
       lightgbmlib.delete_int64_tp(scoredDataLengthLongPtr)
     if (scoredDataOutPtr == null)
```

Comment:
hmm was that a typo on my part, shouldn't the check be != null here...

Reply:
pretty sure that was a typo. so this and the check below (if (leafIndexDataOutPtr == null)) should be changed to !=.

Reply:
It is fixed.